[R4R] ALM

jochym opened the pull request:
Python module version of Alamode program (https://github.com/ttadano/alamode). It is based on my fork of the code for now. It will be transfered to the upstream source at the later time.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/alm```) and found it was in an excellent condition.

jochym commented:
New datapoint - it builds and runs fine on `defaults, conda-forge` channels but fails if the order is reversed.
This seems to be a bug in the system? @conda-forge/core I hate to take your time guys but I at the end of my rope here.

isuruf commented:
@jochym, this is a bug in the package. You need to add `-fopenmp` to `LDFLAGS`

jochym commented:
You mean in the build.sh, before calling setup ? I'll try, but `-fopenmp` is added to compiler flags in the setup.py already.

jochym commented:
@isuruf please take a look now. The missing symbol changed from GOMP_critical_end to dgeqp3_ (which is in lapack). Any ideas?

jochym commented:
Ok. Do not bother @isuruf  - I think it is actually a problem with ALM sources.

jochym commented:
@conda-forge/core I think some work on azure pipelines (https://github.com/conda-forge/staged-recipes/pull/7085 ?	@mariusvniekerk ?) may have broken staged-recipes. The `Unable to find '/azure-pipelines.yml'` shows up suddenly not only in my PR but also elsewere (https://github.com/conda-forge/staged-recipes/pull/7086 , https://github.com/conda-forge/staged-recipes/pull/7080 etc).

jochym commented:
@conda-forge-admin, please rerender

jochym commented:
Thanks! How should I make the PR green again @mariusvniekerk ? It is ready for review and I do not need to add anything more to the rcp.

jochym commented:
@isuruf I think this is ready for review/merging. I have added better tests and cleaned up the package a bit.
I have also performed extensive tests on my production system of the package build from this recipe locally. It seems to be working fine.

jochym commented:
Would you look at this feedstock @jakirkham ?

sodre commented on recipes/alm/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+    - eigen
+  host:
+    - python
+    - blas 1.1 openblas  # [unix]
```
I think what you are looking for here is `- blas * {{ blas_impl }}`. Furthermore, the '# [unix]' tag is redundant because of the `skip` line.

If it does not build, please revert and consider this comment closed.

jochym replied:
Where is a doc about it. I have tried `- blas * {{ openblas }}` but this did not build. Should it be `- blas 1.1 {{ openblas }}` ?

sodre replied:
https://conda-forge.org/docs/blas.html

jochym replied:
But that talks only on conda install forms not the meta.yaml syntax? These are fairly standard packages. What I need is info on what the package which needs blas/lapack should use if it needs any implementation of them and what if it needs specyfic variant (e.g. openblas).

sodre commented on recipes/alm/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+    - setuptools
+  run:
+    - python
+    - blas 1.1 openblas  # [unix]
```
same as in line L26

sodre commented on recipes/alm/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+
+extra:
+  recipe-maintainers:
+    - jochym
```
@ttadano, it looked like you were happy to have your package added to conda-forge. Would you like to be added as a co-maintainer?

ttadano replied:
Yes. @jochym, could you add me to the list?

sodre commented on recipes/alm/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+    - python
+    - blas 1.1 openblas  # [unix]
+    - {{ pin_compatible('numpy') }}
+    - spglib >=1.11.1.2
```
This dependency is too lax and can cause the software to break if the ABI of spglib changes. 

There are two way to fix this, 
  1- In the spglib-feedstock add a `run_exports` section and then remove run dependency from here, see [conda-forge doc](https://conda-forge.org/docs/meta.html#using-conda-build-s-run-exports-feature) for more details.
  2- Use `pin_compatible` just as you did for NumPy, see [conda's doc](https://conda.io/docs/user-guide/tasks/build-packages/variants.html#pinning-at-the-recipe-level) for more details.

Since you are the maintainer for spglib, my suggestion is to fix the spglib recipe and not have to worry about it again for every package that depends on it.

jochym replied:
Opted for the first solution

sodre commented on recipes/alm/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+{% set name = "ALM" %}
+{% set version = "1.0.2" %}
+{% set sha256 = "a3ee1b25a897ec8bd576dd00c948657c4faefbd42f2616bf5ad2ce79a3d8cda4" %}
```
Move the sha256 to L11 for consistency with the new example recipe.

jochym replied:
Done

jochym commented:
Thanks for the review @sodre ! I'll work on it shortly.

jochym commented:
@conda-forge/staged-recipes Please take a look. It seems to be something wrong with this and it is getting worse not better with time. Any assistance will be appreciated.

mariusvniekerk commented:
Rebase on the latest master

jochym commented:
@conda-forge-admin, please rerender

jochym commented:
@mariusvniekerk I am still unable to properly link to openblas. I can build working conda package for ALM using standard conda-build. Somehow I cannot do it for conda-forge, Could you help or point me in the correct direction?

isuruf commented:
See https://conda-forge.org/docs/maintainer/knowledge_base.html#blas

jochym commented:
Thanks. That is what I needed!

isuruf commented on recipes/alm/build.sh:
```diff
@@ -0,0 +1,19 @@
+
+echo $OSX_ARCH
+if [ ${OSX_ARCH}. == . ]; then
+    export LDFLAGS="-fopenmp -lopenblas"    
```
You need to change this as well.

jochym replied:
Thanks!

jochym commented on recipes/alm/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+  version: {{ version }}
+
+source:
+  url: https://github.com/jochym/{{ name }}/archive/v{{ version }}.tar.gz
```
Unfortunately not possible yet. The upstream does not use packages and/or tags yet. I effectively just tag the release on my tree to maka a tarball for packaging. I understand that we should not build from the repo.

jochym commented:
I think, I am done. The windows build is not supported upstream for now. I think it should be possible but first some dependencies need to be ported as well.

isuruf commented:
@jochym, can you send a separate PR with just your changes? It's hard to review this recipe with all the other diff.

jochym commented:
OK. This was a long and twisty road. I will prepare a new PR with just the final recipe. But in the morning - it is 2:30 am here. I will ping you. Thanks for your help.

jakirkham commented:
To help move things along, I took the liberty of just merging `master` into this PR. That seems to have cleaned up things nicely. @isuruf, do you mind taking another look when you have a chance? 🙂

isuruf commented on recipes/alm/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - eigen    
```
Can you remove this?

jochym replied:
I am not sure I follow. eigen seems to be required for build (it is not enough to have it in host), and cxx is required for the c++ extension part of the package. But i will try if it builds without it.

jochym replied:
OK. It seems eigen is not required any more in build. So I am dropping it. Compiler of course is.

jochym commented:
I think this is ready to merge. Please review again and thank you @isuruf  @sodre @jakirkham @mariusvniekerk !
